docs/book/beyond_basics/using_finalizers.md

```diff
@@ -0,0 +1,99 @@
+# Using Finalizers
+
+`Finalizers` allow controllers to implement asynchronous pre-delete hooks. Let
+say you create an external resource such as a storage bucket for each object of
+the API type you are implementing and you would like to clean up the external resource
+when the corresponding object is deleted from Kubernetes, you can use a
+finalizer to delete the external resource.
+
+You can read more about the finalizers in the [kubernetes reference docs](https://kubernetes.io/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions/#finalizers). Section below
+demonstrates how to register and trigger pre-delete hooks in the `Reconcile`
+method of a controller.
+
+{% method %}
+
+Highlights:
+- If object is not being deleted and does not have the finalizer registered,
+  then add the finalizer and update the object in kubernetes.
+- If object is being deleted and the finalizer is still present in finalizers list,
+  then execute the pre-delete logic and remove the finalizer and update the
+  object.
+- You should implement the pre-delete logic in such a way that it is safe to 
+ invoke it multiple times for the same object.
+
+
+```go
+
+func (r *Reconciler) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	err := r.Get(context.TODO(), request.NamespacedName, instance)
+	if err != nil {
+		// handle error
+	}
+
+	// name of your custom finalizer
+	myFinalizerName := "storage.finalizers.example.com"
+
+	if instance.ObjectMeta.DeletionTimestamp.IsZero() {
+		// The object is not being deleted, so if it does not have our finalizer,
+		// then lets add the finalizer and update the object.
+		if !containsString(instance.ObjectMeta.Finalizers, myFinalizerName) {
+			instance.ObjectMeta.Finalizers = append(instance.ObjectMeta.Finalizers, myFinalizerName)
+			if err := r.Update(context.Background(), instance); err != nil {
+				return reconcile.Result{}, err
```

Comment:
Should we return?  Should we re-enqueue?

Reply:
Good catch, it should be re-queue.

Reply:
Done.